sys/arm64/vmm/vmm_dev.c

```diff
@@ -0,0 +1,1041 @@
+		if (error)
+			goto done;
+	}
+
+	error = vm_alloc_memseg(sc->vm, mseg->segid, mseg->len, sysmem);
+	if (error)
+		goto done;
+
+	if (VM_MEMSEG_NAME(mseg)) {
+		error = devmem_create_cdev(vm_name(sc->vm), mseg->segid, name);
+		if (error)
+			vm_free_memseg(sc->vm, mseg->segid);
+		else
+			name = NULL;	/* freed when 'cdev' is destroyed */
+	}
+done:
+	free(name, M_VMMDEV);
+	return (error);
+}
+
+static int
+vm_get_register_set(struct vcpu *vcpu, unsigned int count, int *regnum,
+    uintcap_t *regval)
+{
+	int error, i;
+
+	error = 0;
+	for (i = 0; i < count; i++) {
+		error = vm_get_register(vcpu, regnum[i], &regval[i]);
+		if (error)
+			break;
+	}
+	return (error);
+}
+
+static int
+vm_set_register_set(struct vcpu *vcpu, unsigned int count, int *regnum,
+    uintcap_t *regval)
+{
+	int error, i;
+
+	error = 0;
+	for (i = 0; i < count; i++) {
+		error = vm_set_register(vcpu, regnum[i], regval[i]);
+		if (error)
+			break;
+	}
+	return (error);
+}
+
+static int
+vmmdev_ioctl(struct cdev *cdev, u_long cmd, caddr_t data, int fflag,
+	     struct thread *td)
+{
+	int error, vcpuid, size;
+	cpuset_t *cpuset;
+	struct vmmdev_softc *sc;
+	struct vcpu *vcpu;
+	struct vm_register *vmreg;
+	struct vm_register_set *vmregset;
+	struct vm_run *vmrun;
+	struct vm_vgic_version *vgv;
+	struct vm_vgic_descr *vgic;
+	struct vm_cpuset *vm_cpuset;
+	struct vm_irq *vi;
+	struct vm_capability *vmcap;
+	struct vm_stats *vmstats;
+	struct vm_stat_desc *statdesc;
+	struct vm_suspend *vmsuspend;
+	struct vm_exception *vmexc;
+	struct vm_gla2gpa *gg;
+	struct vm_memmap *mm;
+	struct vm_msi *vmsi;
+	struct vm_cpu_topology *topology;
+	uintcap_t *regvals;
+	int *regnums;
+	enum { NONE, SINGLE, ALL } vcpus_locked;
+	bool memsegs_locked;
+
+	error = vmm_priv_check(curthread->td_ucred);
+	if (error)
+		return (error);
+
+	sc = vmmdev_lookup2(cdev);
+	if (sc == NULL)
+		return (ENXIO);
+
+	error = 0;
+	vcpuid = -1;
+	vcpu = NULL;
+	vcpus_locked = NONE;
+	memsegs_locked = false;
+
+	/*
+	 * Some VMM ioctls can operate only on vcpus that are not running.
+	 */
+	switch (cmd) {
+	case VM_RUN:
+	case VM_GET_REGISTER:
+	case VM_SET_REGISTER:
+	case VM_GET_REGISTER_SET:
+	case VM_SET_REGISTER_SET:
+	case VM_INJECT_EXCEPTION:
+	case VM_GET_CAPABILITY:
+	case VM_SET_CAPABILITY:
+	case VM_GLA2GPA_NOFAULT:
+	case VM_ACTIVATE_CPU:
+		/*
+		 * ioctls that can operate only on vcpus that are not running.
+		 */
+		vcpuid = *(int *)data;
+		vcpu = vm_alloc_vcpu(sc->vm, vcpuid);
+		if (vcpu == NULL) {
+			error = EINVAL;
+			goto done;
+		}
+		error = vcpu_lock_one(vcpu);
+		if (error)
+			goto done;
+		vcpus_locked = SINGLE;
+		break;
+
+	case VM_ALLOC_MEMSEG:
+	case VM_MMAP_MEMSEG:
+	/* TODO: VM_MUNMAP_MEMSEG */
+	case VM_REINIT:
+	case VM_ATTACH_VGIC:
+		/*
+		 * ioctls that modify the memory map must lock memory
+		 * segments exclusively.
+		 */
+		vm_xlock_memsegs(sc->vm);
+		memsegs_locked = true;
+		/* FALLTHROUGH */
+	/* TODO: Any non-memseg ioctls? */
```

Comment:
FALLTHROUGH but no case statement?

Reply:
Yes, this is a bit gross. I'll get this fixed.